kyo-core/shared/src/main/scala/kyo/envs.scala

```diff
@@ -3,45 +3,61 @@ package kyo
 import kyo.core.*
 
 class Envs[+V] extends Effect[Envs[V]]:
-    type Command[T] = Unit
+    type Command[T] = Tag[T]
 
 object Envs:
     private case object envs extends Envs[Any]
     private def envs[V]: Envs[V] = envs.asInstanceOf[Envs[V]]
 
-    def get[V](using Tag[Envs[V]]): V < Envs[V] =
-        envs[V].suspend[V](())
+    // Envs.get[String] would normally use a tag of Envs[String]
+    // Yet we may provide a Layer with an output of String & Int.
+    // This would usually give the handler a tag of Envs[String & Int]
+    // However, then handler.tag != kyo.tag in the handle implementation.
+    // So we have to use this type erased tag and rely on our `accepts` implementation
+    private trait EnvsErased
+    private val envsTag: Tag[EnvsErased]   = Tag[EnvsErased]
+    private inline given [V]: Tag[Envs[V]] = envsTag.asInstanceOf[Tag[Envs[V]]]
+
+    def get[V](using tag: Tag[V]): V < Envs[V] =
+        envs[V].suspend[V](tag)
+
+    inline def run[V: Tag, T: Flat, S, VS, VR](env: V)(value: T < (Envs[VS] & S))(
+        using HasEnvs[V, VS] { type Remainder = VR }
+    ): T < (S & VR) =
+        envs[V].handle(handler[V])(TypeMap(env), value).asInstanceOf[T < (S & VR)]
```

Comment:
I'd prefer to avoid the `TypeMap` map allocation. Could this method use a dedicated handler?

Reply:
Hm. I just realized, then we might need to allocate a tuple for the state:

```scala
    private val cachedValueHandler =
        new ResultHandler[Tag[Any], Tag, Envs[Any], Id, Any]:
            override def accepts[T](st: (Any, Tag[Any]), command: Tag[T]): Boolean =
                st._2 <:< command // <--- HERE.      
```

We need both the value and a way to match on the command, because when we suspend, we're using the Tag[ErasedEnv], because we don't know which handler will eventually run the suspension. Is it still worth doing this if there's an allocation either way?

Reply:
Oh, I see. I think we could change `core.handle` to pass the tag to `accepts` so we can avoid having to hold the tag in the state.

Reply:
we could do that as a follow up if you like

Reply:
Ah, I think the problem is that the Effect tag needs to be `ErasedEnv` and here:
```scala
            override def accepts[T](st: TypeMap[Any], command: Tag[T]): Boolean =
                st <:< command
```

The TypeMap is both the State and ALSO can be used to check if it contains the Tag[Value] for the suspended Envs.get[Value].

But for the plain value handler, the state would simply be the Value.
```scala
            override def accepts[T](st: Any, command: Tag[T]): Boolean =
                ???
```

And we'd have no way of seeing if should handle this Tag. The tag of the effect is still going to be Tag[ErasedEnv], so that wouldn't help us in this case, even if it got passed in.

Am I making sense? I could also be confused 😅

Reply:
Couldn't we use the regular `Tag[Envs[V]]` instead of `EnvsErased` in this method?

Reply:
btw, if this is getting complicated, I'm ok with merging with the allocation! We can address it later.

Reply:
The problem is that we need to use the same tag in the suspension _and_ the handler. And when we suspend, we won't know which handler is going to handle it :)

Reply:
But definitely worth noting and possibly re-examining later. It'd certainly be nice to avoid. State could also be a union of `Tag[?] | TypeMap[?]` ? But that would add some cost to the lookup.